heat/core/_operations.py

```diff
@@ -62,194 +62,202 @@ def __binary_op(
-                    # color = 0 if t1.comm.rank < t2.shape[t1.split] else 1
-                    # newcomm = t1.comm.Split(color, t1.comm.rank)
-                    # if t1.comm.rank > 0 and color == 0:
-                    #     t1.larray = torch.zeros(
-                    #         t1.shape, dtype=t1.dtype.torch_type(), device=t1.device.torch_device
-                    #     )
-                    # newcomm.Bcast(t1)
-                    # newcomm.Free()
-                    # t1.__lshape = t1.gshape
-                    # t1.__split = None
-                    # t1.__lshape_map = torch.tensor(t1.gshape).unsqueeze_(0)
-                    # t1.__balanced = True
-            if t2.split is not None:
-                if t2.shape[t2.split] == 1 and t2.comm.is_distributed():
-                    # warnings.warn(
-                    #     "Broadcasting requires transferring data of second operator between MPI ranks!"
-                    # )
-                    t2.resplit_(None)
-                    # color = 0 if t2.comm.rank < t1.shape[t2.split] else 1
-                    # newcomm = t2.comm.Split(color, t2.comm.rank)
-                    # if t2.comm.rank > 0 and color == 0:
-                    #     t2.larray = torch.zeros(
-                    #         t2.shape, dtype=t2.dtype.torch_type(), device=t2.device.torch_device
-                    #     )
-                    # newcomm.Bcast(t2)
-                    # newcomm.Free()
-                    # t2.__lshape = t2.gshape
-                    # t2.__split = None
-                    # t2.__lshape_map = torch.tensor(t2.gshape).unsqueeze_(0)
-                    # t2.__balanced = True
-        else:
-            raise TypeError(
-                "Only tensors and numeric scalars are supported, but input was {}".format(type(t2))
-            )
-    else:
-        raise NotImplementedError("Not implemented for non scalar")
-
-    if t1.split is not None:
+    elif np.isscalar(t1) and isinstance(t2, DNDarray):
+        try:
+            t1 = factories.array(t1, device=t2.device, comm=t2.comm)
+        except (ValueError, TypeError):
+            raise TypeError("Data type not supported, input was {}".format(type(t1)))
+    elif isinstance(t1, DNDarray) and np.isscalar(t2):
+        try:
+            t2 = factories.array(t2, device=t1.device, comm=t1.comm)
+        except (ValueError, TypeError):
+            raise TypeError("Data type not supported, input was {}".format(type(t2)))
+    sanitation.sanitize_in(t1)
+    sanitation.sanitize_in(t2)
+
+    # Make inputs have the same dimensionality
+    output_shape = stride_tricks.broadcast_shape(t1.shape, t2.shape)
+    # Broadcasting allows additional empty dimensions on the left side
+    # TODO simplify this once newaxis-indexing is supported to get rid of the loops
+    while len(t1.shape) < len(output_shape):
+        t1 = t1.expand_dims(axis=0)
+    while len(t2.shape) < len(output_shape):
+        t2 = t2.expand_dims(axis=0)
```

Comment:
np.expand_dims() allows tuples for 'axis'. We will have to improve our implementation of this function.